### PR TITLE
Tool use and model name for local models

### PR DESCRIPTION
**Pull request by Jeadie**
## 🗣 Description
 - Ensure the model name from spicepod is returned in the LLM response.
 - Pass all chat/completion stream request parameters to model pipeline. This requires a refactor of functions. 
 - Ensure we return `tool_calls` finish reason when appropriate. 
 - Fix `created` response. It should be seconds, but mistralrs uses milliseconds.

### CODE REVIEW

File: crates/llms/src/chat/mistral.rs [hunk 1]
```diff
@@ -356,38 +386,44 @@ impl MistralLlama {
...
+    /// Process a single `stream=false` request to generate `OpenAi` chat completion.
+    fn post_process_req(resp: MistralResponse) -> Result<ChatCompletionResponse> {
+        match resp {
+            MistralResponse::Done(mut resp) => {
+                // mistralrs does not return "tool_calls" as a finish_reason correctly (like OpenAI spec).
+                // This is a workaround to set it correctly.
+                resp.choices.iter_mut().for_each(|c| {
+                    if c.finish_reason == "stop" && !c.message.tool_calls.is_empty() {
+                        c.finish_reason = "tool_calls".to_string();
+                    }
```
Comment (Sevenannn) [inline crates/llms/src/chat/mistral.rs:408]: Is the `finish_reason==stop` and `!tool_calls.is_empty()` sufficient to prove that the `finish_reason` is actually tool_call, could the finish reason something else (e.g. content filter) when both requirements are meet in this case.

File: crates/llms/src/chat/mistral.rs [hunk 2]
```diff
@@ -340,14 +343,41 @@ impl MistralLlama {
         })
     }
 
-    async fn run_internal(
+    /// Prepares and sends a [`CreateChatCompletionRequest`] to the model pipeline.
+    #[allow(deprecated, clippy::cast_possible_truncation)]
```
Comment (sgrebnov) [inline crates/llms/src/chat/mistral.rs]: @Jeadie  - could you please add a note why `deprecated` api is used,

File: crates/llms/src/chat/mistral.rs [hunk 3]
```diff
@@ -340,14 +343,41 @@ impl MistralLlama {
...
+                Stop::String(s) => mistralrs::StopTokens::Seqs(vec![s]),
+                Stop::StringArray(s) => mistralrs::StopTokens::Seqs(s),
+            }),
+            max_len: req.max_completion_tokens.map(|x| x as usize),
+            // logits_bias: req.logit_bias,
+            logits_bias: None,
+            n_choices: req.n.unwrap_or(1) as usize,
+            dry_params: None,
+        };
+        let (snd, rcv) = channel::<MistralResponse>(10_000);
```
Comment (sgrebnov) [inline crates/llms/src/chat/mistral.rs]: (tx, rx) is a more common naming pattern

File: crates/llms/src/chat/mistral.rs [hunk 4]
```diff
@@ -407,141 +443,96 @@ impl Chat for MistralLlama {
...
+                r#type: None,
+                param: None,
+                code: None,
+            }));
         };
 
-        let resp = self
-            .run_internal(messages, tools, tool_choice, Some(p))
-            .await
+        let z = Self::post_process_req(resp)
```
Comment (sgrebnov) [inline crates/llms/src/chat/mistral.rs]: Naming: `z` is hard to read/understand the purpose, please see if there could be a better alternative

**Review by sgrebnov — DISMISSED** [review]

File: crates/llms/src/chat/mistral.rs [hunk 5]
```diff
@@ -340,14 +343,41 @@ impl MistralLlama {
...
+            min_p: None,
+            top_n_logprobs: req.top_logprobs.unwrap_or_default().into(),
+            frequency_penalty: req.frequency_penalty,
+            presence_penalty: req.presence_penalty,
+            stop_toks: req.stop.map(|s| match s {
+                Stop::String(s) => mistralrs::StopTokens::Seqs(vec![s]),
+                Stop::StringArray(s) => mistralrs::StopTokens::Seqs(s),
+            }),
+            max_len: req.max_completion_tokens.map(|x| x as usize),
+            // logits_bias: req.logit_bias,
```
Comment (phillipleblanc) [inline crates/llms/src/chat/mistral.rs]:
```suggestion
```

File: crates/llms/src/chat/mistral.rs [hunk 6]
```diff
@@ -562,6 +553,102 @@ impl Chat for MistralLlama {
     }
 }
 
+fn stream_from_response(
+    mut rcv: Receiver<MistralResponse>,
+) -> Pin<Box<dyn Stream<Item = Result<CreateChatCompletionStreamResponse, OpenAIError>> + Send>> {
+    Pin::from(Box::new(stream! {
```
Comment (phillipleblanc) [inline crates/llms/src/chat/mistral.rs]:
```suggestion
    Box::pin(stream! {
```

File: crates/llms/src/chat/mistral.rs (hunk 1 above)
Comment (Jeadie) [reply crates/llms/src/chat/mistral.rs:408]: no, ` !tool_calls.is_empty(` is pretty definitive in my experience. There are other stop reasons for these cases `ContentFilter` or `Stop` if `max_completion_tokens` is reached.

**Review by sgrebnov — APPROVED** [review]

**Pull request merged**